File: samples/common/python/trt_utils.py
# SPDX-FileCopyrightText: Copyright (c) 2022-2023 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
trt_utils

This file hosts various TensorRT related utilities.
"""

import logging
import torch
import numpy as np
import tensorrt as trt


def convert_onnx_to_tensorrt(
    onnx_file_path,
    trt_engine_file_path,
    max_batch_size,
    max_workspace_size=5,
):
    """
    Converts an ONNX engine to a serialized TensorRT engine.
    :param onnx_file_path: Full path to an existing ONNX file.
    :param trt_engine_file_path: Full path to save the generated TensorRT Engine file.
    :param max_batch_size: The maximum batch size to use in the TensorRT engine.
    :param max_workspace_size: The maximum GPU memory that TensorRT can use (in GB.)
    :return: True if engine was generated. False otherwise.
    """
    current_log_level = logging.root.level
    if current_log_level == logging.INFO:
        trt_log_level = trt.Logger.INFO
    elif current_log_level == logging.ERROR:
        trt_log_level = trt.Logger.ERROR
    elif current_log_level == logging.WARNING:
        trt_log_level = trt.Logger.WARNING
    elif current_log_level == logging.DEBUG:
        trt_log_level = trt.Logger.VERBOSE
    else:
        trt_log_level = trt.Logger.INFO

    trt_logger = trt.Logger(trt_log_level)
    trt_logger.log(trt.Logger.INFO, "Using TensorRT version: %s" % trt.__version__)

    with trt.Builder(trt_logger) as builder, builder.create_network(
        1 << int(trt.NetworkDefinitionCreationFlag.EXPLICIT_BATCH)
    ) as network, trt.OnnxParser(network, trt_logger) as parser:

        config = builder.create_builder_config()
        config.set_memory_pool_limit(
            trt.MemoryPoolType.WORKSPACE, 1024 * 1024 * 1024 * max_workspace_size
        )  # Sets workspace size in GB.
        if builder.platform_has_fast_fp16:
            config.set_flag(trt.BuilderFlag.FP16)
            config.set_flag(trt.BuilderFlag.STRICT_TYPES)
            trt_logger.log(trt.Logger.INFO, "Using precision : float16")
        else:
            trt_logger.log(trt.Logger.INFO, "Using precision : float32")

        # Parse model file
        trt_logger.log(
            trt.Logger.INFO, "Loading ONNX file from path %s " % onnx_file_path
        )
        with open(onnx_file_path, "rb") as model:
            if not parser.parse(model.read()):
                raise ValueError("Failed to parse the ONNX engine.")

        inputs = [network.get_input(i) for i in range(network.num_inputs)]
        profile = builder.create_optimization_profile()
        dynamic_inputs = False
        for input_tensor in inputs:
            if input_tensor.shape[0] == -1:
                dynamic_inputs = True
                min_shape = [1] + list(input_tensor.shape[1:])
                opt_shape = [max_batch_size] + list(input_tensor.shape[1:])
                max_shape = [max_batch_size] + list(input_tensor.shape[1:])
                profile.set_shape(input_tensor.name, min_shape, opt_shape, max_shape)
        if dynamic_inputs:
            config.add_optimization_profile(profile)

        engine = builder.build_serialized_network(network, config)

        if not engine:
            raise ValueError("Failed to generate the TensorRT engine.")

        with open(trt_engine_file_path, "wb") as f:
            f.write(engine)
            trt_logger.log(
                trt.Logger.INFO, "Wrote TensorRT engine file: %s" % trt_engine_file_path
            )


def setup_tensort_bindings(trt_model, batch_size, device_id, logger):
    """
    Setups the I/O bindings for a TensorRT engine for the first time.
    :param trt_model: Full path to the generated TensorRT Engine file.
    :param batch_size: The maximum batch size that should be supported in the model.
    :param device_id: The GPU device id on which you want to allocated the buffers.
    :return: A list of output tensors and the index of the first output.
    """
    # For TensorRT, we need to allocate the output data buffers.
    # The input data buffers are already allocated by us.
    output_binding_idx = 0
    output_idx = 0
    output_tensors = []

    # Loop over all the I/O bindings.
    for b_idx in range(trt_model.num_io_tensors):
        # Get various properties associated with the bindings.
        b_name = trt_model.get_tensor_name(b_idx)
        b_shape = tuple(trt_model.get_tensor_shape(b_name))
        b_dtype = np.dtype(trt.nptype(trt_model.get_tensor_dtype(b_name))).name

        # Append to the appropriate list.
        if trt_model.get_tensor_mode(b_name) == trt.TensorIOMode.OUTPUT:
            # First allocate on device output buffers, using PyTorch.
            # Get the C, H, W dimensions from the layer shape to set the output layer size for the buffer
            output = torch.zeros(
                size=(batch_size, b_shape[-3], b_shape[-2], b_shape[-1]),
                dtype=getattr(torch, b_dtype),
                device="cuda:%d" % device_id,
            )
            # Since we know the name of our output layer, we will check against
            # it and grab its binding index.
            if b_name == "output":
                output_idx = output_binding_idx

            output_binding_idx += 1
            output_tensors.append(output)

    return output_tensors, output_idx

File: samples/common/TRTUtils.cpp
/*
 * SPDX-FileCopyrightText: Copyright (c) 2022 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "TRTUtils.h"

#include "NvInfer.h"
#include "NvUtils.h"

#include <algorithm>
#include <fstream>
#include <iostream>
#include <stdexcept>

class TRTLogger : public nvinfer1::ILogger
{
public:
    void log(nvinfer1::ILogger::Severity severity, const char *msg) noexcept override
    {
        switch (severity)
        {
        case nvinfer1::ILogger::Severity::kINTERNAL_ERROR:
        {
            std::cout << msg << std::endl;
            break;
        }
        case nvinfer1::ILogger::Severity::kERROR:
        {
            std::cout << msg << std::endl;
            break;
        }
        default:
        {
            break;
        }
        }
    }
};

struct TRTBackend::TRTImpl
{
    TRTImpl()
        : m_logger(new TRTLogger())
        , m_TRTRuntime(nullptr, [](nvinfer1::IRuntime *runtime) { runtime->destroy(); })
        , m_inferenceEngine(nullptr)
        , m_ownedInferenceEngine(nullptr, [](nvinfer1::ICudaEngine *eng) { eng->destroy(); })
        , m_inferContext(nullptr, [](nvinfer1::IExecutionContext *ectx) { ectx->destroy(); })
        , m_cudaStream(0)
    {
    }

    std::unique_ptr<TRTLogger>                                                            m_logger;
    std::unique_ptr<nvinfer1::IRuntime, void (*)(nvinfer1::IRuntime *)>                   m_TRTRuntime;
    nvinfer1::ICudaEngine                                                                *m_inferenceEngine;
    std::unique_ptr<nvinfer1::ICudaEngine, void (*)(nvinfer1::ICudaEngine *)>             m_ownedInferenceEngine;
    std::unique_ptr<nvinfer1::IExecutionContext, void (*)(nvinfer1::IExecutionContext *)> m_inferContext;

    cudaStream_t                         m_cudaStream;
    int                                  m_bindingsCount = 0;
    int                                  m_batchSize     = 1;
    std::unordered_map<std::string, int> m_blobMap;

    void loadNetWorkFromFile(const char *modelFilePath);
    // find the input/output bindings
    void setupIO(int batchSize);
};

void TRTBackend::TRTImpl::loadNetWorkFromFile(const char *modelFilePath)
{
    // Initialize TRT engine and deserialize it from file
    std::ifstream           trtModelFStream(modelFilePath, std::ios::binary);
    std::unique_ptr<char[]> trtModelContent;
    size_t                  trtModelContentSize = 0;
    if (!trtModelFStream.good())
    {
        std::cerr << "Model File: " << modelFilePath << std::endl;
        throw std::runtime_error("TensorRT: Model file not found.");
    }
    else
    {
        trtModelFStream.seekg(0, trtModelFStream.end);
        trtModelContentSize = trtModelFStream.tellg();
        trtModelFStream.seekg(0, trtModelFStream.beg);
        trtModelContent.reset(new char[trtModelContentSize]);
        trtModelFStream.read(trtModelContent.get(), trtModelContentSize);
        trtModelFStream.close();
        std::cout << "Deserializing engine from: " << modelFilePath;
    }
    m_TRTRuntime.reset(nvinfer1::createInferRuntime(*(m_logger.get())));
    m_inferenceEngine = dynamic_cast<nvinfer1::ICudaEngine *>(
        m_TRTRuntime->deserializeCudaEngine(trtModelContent.get(), trtModelContentSize, nullptr));
    m_ownedInferenceEngine.reset(m_inferenceEngine);
    m_inferContext.reset(m_inferenceEngine->createExecutionContext());
    m_inferContext->setOptimizationProfile(0);
}

void TRTBackend::TRTImpl::setupIO(int batchSize)
{
    // @TODO: use getBindingDimensions to avoid re-setting the IO.
    m_bindingsCount = m_inferenceEngine->getNbBindings();
    for (int i = 0; i < m_bindingsCount; i++)
    {
        m_blobMap[std::string(m_inferenceEngine->getBindingName(i))] = i;
        if (m_inferenceEngine->bindingIsInput(i))
        {
            nvinfer1::Dims  dims_i(m_inferenceEngine->getBindingDimensions(i));
            nvinfer1::Dims4 inputDims{batchSize, dims_i.d[1], dims_i.d[2], dims_i.d[3]};
            m_inferContext->setBindingDimensions(i, inputDims);
        }
    }
}

TRTBackend::TRTBackend(const char *modelFilePath, int batchSize)
    : m_pImpl(new TRTImpl())
{
    m_pImpl->m_batchSize = batchSize;
    m_pImpl->loadNetWorkFromFile(modelFilePath);
    m_pImpl->setupIO(m_pImpl->m_batchSize);
}

TRTBackend::~TRTBackend() {}

void TRTBackend::infer(void **buffer, int batchSize, cudaStream_t stream)
{
    m_pImpl->setupIO(batchSize);

    bool success = true;
    if (!m_pImpl->m_inferenceEngine->hasImplicitBatchDimension())
    {
        m_pImpl->m_inferContext->enqueueV2(buffer, stream, nullptr);
    }
    else
    {
        m_pImpl->m_inferContext->enqueue(batchSize, buffer, stream, nullptr);
    }

    if (!success)
    {
        throw std::runtime_error("TensorRT: Inference failed");
    }
}

int TRTBackend::getBlobCount() const
{
    return m_pImpl->m_bindingsCount;
}

TRTBackendBlobSize TRTBackend::getTRTBackendBlobSize(int blobIndex) const
{
    if (blobIndex >= m_pImpl->m_bindingsCount)
    {
        throw std::runtime_error("blobIndex out of range");
    }
    auto               dim = m_pImpl->m_inferenceEngine->getBindingDimensions(blobIndex);
    TRTBackendBlobSize blobSize;
    if (dim.nbDims == 2)
    {
        blobSize = {1, dim.d[0], dim.d[1]};
    }
    else if (dim.nbDims == 3)
    {
        blobSize = {dim.d[0], dim.d[1], dim.d[2]};
    }
    else if (dim.nbDims == 4)
    {
        blobSize = {dim.d[1], dim.d[2], dim.d[3]};
    }
    else
    {
        throw std::runtime_error("Unknown TensorRT binding dimension!");
    }
    return blobSize;
}

int TRTBackend::getBlobLinearSize(int blobIndex) const
{
    const TRTBackendBlobSize shape = getTRTBackendBlobSize(blobIndex);
    nvinfer1::Dims3          dims_val{shape.channels, shape.height, shape.width};
    int                      blobSize = 1;
    for (int i = 0; i < 3; i++)
    {
        blobSize *= dims_val.d[i] <= 0 ? 1 : dims_val.d[i];
    }
    return blobSize;
}

int TRTBackend::getBlobIndex(const char *blobName) const
{
    auto blobItr = m_pImpl->m_blobMap.find(std::string(blobName));
    if (blobItr == m_pImpl->m_blobMap.end())
    {
        throw std::runtime_error("blobName not found");
    }
    return blobItr->second;
}

bool TRTBackend::bindingIsInput(const int index) const
{
    return m_pImpl->m_inferenceEngine->bindingIsInput(index);
}

const char *TRTBackend::getLayerName(const int index) const
{
    return m_pImpl->m_ownedInferenceEngine->getBindingName(index);
}
